[Azure.Core] Update ReadMe for ModelSerializer (#37998)
Azure/azure-sdk-for-net: sdk/core/Azure.Core/samples/Serialization.md

# Azure.Core public serialization

> The .NET Azure SDK libraries support serialization and deserialization of most client library model types using the static `ModelSerializer` class. 

Reviewer: Adding https://github.com/Azure/azure-sdk-for-net/pull/37998/files/5c598cdd9acecca64157142bcd5e1a9ed9abc008#r1283441598 back, as it looks like it was lost in an edit.

## Key Concepts

- [Using the ModelSerializer](#using-the-modelserializer)
- [Using models with protocol methods](#using-models-with-protocol-methods)

> - [Using JsonSerializer](#using-jsonserializer)

Reviewer 2: The title for this could be confusing.  You're telling me in the overview that `ModelSerializer` is the thing to use for client library serialization and then one of the first three key points is talking about another serializer.   

What's the scenario where I'd use `JsonSerializer`?  I'm thinking we should frame it up in that context.  If we don't have a clear and compelling majority champion scenario, maybe we should reconsider whether it should be this prominent and we move it to an appending or other corner-cases area.

- [Envelope bring your own model case](#envelope-bring-your-own-model-case)

## Using the ModelSerializer

> The `ModelSerializer` class enables users to serialize and deserialize any Azure models that implement the `IModelSerializable<T>` interface. The following example shows a Dog model being serialized and deserialized.

Reviewer 2: The first question that comes to mind after reading this is "how do I know if a model supports the interface?"  That's followed closely by "why would some have it an others may not?"

```C# Snippet:BaseModelSerializer
Dog doggo = new Dog
{
    Name = "Doggo",
    Age = 7
};

BinaryData data = ModelSerializer.Serialize(doggo);

Dog dog = ModelSerializer.Deserialize<Dog>(data);
```

> By default, the ModelSerializer gives you the Json representation of the model. Some services accept XML so if you want to send the serialized data to an Azure service that accepts XML, you can use the `ModelSerializerFormat` enum to specify the format. By using the `Wire` format, the serializer automatically picks what the service would use.

Reviewer 2: Is `Wire` only useful for services that use XML, or is there a difference between the JSON format as well?  

For example, if I don't use `Wire` and I take my round-trip model, serialize it, store it for a bit, and then send it to my UI in the browser with the JS client library submitting it to the service - does that work or do I need to ensure `Wire`?

What is the "other" format that we use as the default?  Should we touch briefly on which scenarios each is intended to be used for?

```C# Snippet:ModelSerializerWithFormat

> Dog doggo = new Dog

Reviewer 2: Krzysztof left a comment earlier that I agree with, but I think it disappeared.   We should consider using an actual model from one of the libraries - or at least a mock type that strongly resembles one.   `Dog` feels arbitrary and we don't define it anywhere, so I could not just copy/paste this snippet and see it work.

{
    Name = "Doggo",
    Age = 7
};

> ModelSerializerOptions options = new ModelSerializerOptions(format: ModelSerializerFormat.Wire);

Reviewer 2: Moot because we don't actually define `Dog` - but it sure would be confusing to me learning about this if I change to `Wire` format and the output looks exactly the same.   

Some additional discussion of what each format is and when they should be used may help, but we can't guarantee that developers will read it.   The snippets should work copy/paste and demonstrate.

BinaryData data = ModelSerializer.Serialize(doggo, options);

Dog dog = ModelSerializer.Deserialize<Dog>(data, options);
```

> ## Using models with protocol methods

Reviewer 2: To clarify intent here - I recognize that we've chosen to call these "protocol methods" across languages and there is other documentation that discusses them in detail.

What I'm asking for here is just a sentence or maybe two that describes the term for the layperson so that they don't have to jump to another document just to get a definition.  We'll want to keep the link for folks that would like to dig deeper and understand the wider context.

> ## Using models with protocol methods

Reviewer: You can always hyperlink to the [protocol method](https://github.com/Azure/azure-sdk-for-net/blob/main/sdk/core/Azure.Core/samples/ProtocolMethods.md) doc as well.

Dog dog = new Dog
{
    Name = "Doggo",
    Age = 7
};

// Our models contain an implicit cast to RequestContent so you can pass them directly to protocol methods.
Response response = client.CreatePet("myPet", dog);

// Our models also contain an explicit cast to the response type so you can deserialize them easily.
dog = (Dog)response;
```

## Using JsonSerializer

If you would like to use the `JsonSerializer`, you can add the `ModelJsonConverter` to your options, which allow `JsonSerializer` to serialize and deserialize any model that implements `IModelJsonSerializable<T>`. The following example shows a Dog model being serialized and deserialized using the ModelJsonConverter.

```C# Snippet:BaseModelConverter
Dog dog = new Dog
{
    Name = "Doggo",
    Age = 7
};

JsonSerializerOptions options = new JsonSerializerOptions();
// The ModelJsonConverter is able to serialize and deserialize any model that implements IModelJsonSerializable<T>.
options.Converters.Add(new ModelJsonConverter());

string json = JsonSerializer.Serialize(dog, options);

dog = JsonSerializer.Deserialize<Dog>(json, options);
```

## Envelope bring your own model case

> The following examples show a use case where a user brings a model unknown to the Serializer. The serialization used for each model can also be set in the `ModelSerializableOptions` options property `GenericTypeSerializerCreator`. 

Reviewer 3: what are models unknown to the serialize? Do you mean models that don't implement the interface? If yes, I would just say that.

# Azure.Core Public Serialization Guide

> The latest updates to the `Azure.Core` library have simplified serializing and deserializing of all public Azure models. With the addiion of the `ModelSerializer` class, we provide a public interface that exposes the currently internal serialization code so customers can access it without needing to use reflection or write their own translation. The recommended serialization approach is to use the static `ModelSerializer` class. The `ModelSerializer` class also has the `ModelSerializerOptions` which allows the user to set the serialization `ModelSerializerFormat` type (`Json` or `Wire`) and provide custom Serializer types for specific models.

Reviewer 4: ```suggestion
The latest updates to the `Azure.Core` library have simplified serializing and deserializing of all public Azure models. With the addiion of the `ModelSerializer` class, we provide a public interface that exposes the previously internal serialization code so customers can access it without needing to use reflection or write their own translation. The recommended serialization approach is to use the static `ModelSerializer` class. The `ModelSerializer` class also has `ModelSerializerOptions` which allows the user to set the serialization `ModelSerializerFormat` type (`Json` or `Wire`) and provide custom Serializer types for specific models.
```

> The latest updates to the `Azure.Core` library have simplified serializing and deserializing of all public Azure models. With the addiion of the `ModelSerializer` class, we provide a public interface that exposes the currently internal serialization code so customers can access it without needing to use reflection or write their own translation. The recommended serialization approach is to use the static `ModelSerializer` class. The `ModelSerializer` class also has the `ModelSerializerOptions` which allows the user to set the serialization `ModelSerializerFormat` type (`Json` or `Wire`) and provide custom Serializer types for specific models.

Reviewer 3: I would not say "all public Azure models". It might be perceived as overpromising. I would change it "to Azure SDK models"

> The latest updates to the `Azure.Core` library have simplified serializing and deserializing of all public Azure models. With the addiion of the `ModelSerializer` class, we provide a public interface that exposes the currently internal serialization code so customers can access it without needing to use reflection or write their own translation. The recommended serialization approach is to use the static `ModelSerializer` class. The `ModelSerializer` class also has the `ModelSerializerOptions` which allows the user to set the serialization `ModelSerializerFormat` type (`Json` or `Wire`) and provide custom Serializer types for specific models.

Reviewer 3: I also think we need to motivate this feature at the beginning of the document, i.e. why do we need special ModelSerializer and the interfaces? Users will wonder why Azure SDK team is inventing a new serialization technology as opposed to using one of the existing techs.

> The latest updates to the `Azure.Core` library have simplified serializing and deserializing of all public Azure models. With the addiion of the `ModelSerializer` class, we provide a public interface that exposes the currently internal serialization code so customers can access it without needing to use reflection or write their own translation. The recommended serialization approach is to use the static `ModelSerializer` class. The `ModelSerializer` class also has the `ModelSerializerOptions` which allows the user to set the serialization `ModelSerializerFormat` type (`Json` or `Wire`) and provide custom Serializer types for specific models.

Author: KrzysztofCwalina 

This doc should not start with this section. It's impossible to understand it without the context described later in this document.

annelo-msft 

It might be helpful to begin with a short summary of what the interfaces are and how they are intended to be used. For Azure.Core, it can be helpful to think about this from the perspective of an end-user of a library, as well as a library author. Should library authors implement IModelSerializable on all models? Will the generator handle it for them? When do they need to also implement IModelJsonSerializable? What does an end-user need to know they can do with a model that implements these interfaces? Should they understand ModelSerializer and how to use it? When do I need to be thinking about ModelJsonConverter and why? It'd be great to get all these questions answered at the start with some short illustrative samples before diving into further details and all the options types below. That way, when I get to those sections, as a reader I would know what I'm looking at and why it's useful to me.

> ## Key Concepts

Reviewer 4: probably add a section in key concepts around ModelSerializerFormat.  You can start with the descriptions in the enum itself of what the two formats mean.

- [Using the ModelSerializer](#using-the-modelserializer)

> - [Using explicit cast](#using-explicit-cast)

Reviewer 4: Lets call this interfacing with protocol methods since this is the use case where we would expect customers to need the cast.

> - [Using explicit cast](#using-explicit-cast)

Author: KrzysztofCwalina 

there are many things that happen in this sample. I think we should start with a simple sample showing how to serialize/deserialize simple model, only later get into how all of this relates to protocol methods.

> - [Using the ModelJsonConverter](#using-the-modeljsonconverter)

Reviewer 4: Lets call this using JsonSerializer since that is when they would need the ModelJsonConverter

> - [Envelope BYOM Case](#envelope-byom-case)

Reviewer 4: Lets spell this out as simply Bring your own model

> - [Envelope BYOM Case](#envelope-byom-case)

Author: annelo-msft 

This feels like shorthand that people on our team understand that might be hard for users of the interfaces to decode initally. For public-facing documentation, maybe a different title could help readers?

> - [Using ModelSerializer with generic IModelSerializable](#using-modelserializer-with-generic-imodelserializable)

Reviewer 4: Lets calling this interfacing with the low level interfaces

## Using the ModelSerializer

> The default serialization options can be overridden by passing in a `ModelSerializerOptions` object to the Serialize and Deserialize methods. Developers can set the format options to specify the serialization format such as XML, JSON, and BinaryData. Default `ModelSerializationOptions` use the `Wire` format and the `System.Text.Json` serializer. The following samples demonstrate how to use `ModelSerializer` class for `System.Text.Json` and `Newtonsoft.Json` serialization.

Reviewer 4: BinaryData is not an option they can set.

Default uses Json format

> The default serialization options can be overridden by passing in a `ModelSerializerOptions` object to the Serialize and Deserialize methods. Developers can set the format options to specify the serialization format such as XML, JSON, and BinaryData. Default `ModelSerializationOptions` use the `Wire` format and the `System.Text.Json` serializer. The following samples demonstrate how to use `ModelSerializer` class for `System.Text.Json` and `Newtonsoft.Json` serialization.

Reviewer 3: I would not start with describing how to override the defaults. We need to start with describing and illustrating the default behavior and then talk about how to change the defaults.

> The default serialization options can be overridden by passing in a `ModelSerializerOptions` object to the Serialize and Deserialize methods. Developers can set the format options to specify the serialization format such as XML, JSON, and BinaryData. Default `ModelSerializationOptions` use the `Wire` format and the `System.Text.Json` serializer. The following samples demonstrate how to use `ModelSerializer` class for `System.Text.Json` and `Newtonsoft.Json` serialization.

Reviewer 3: BinaryData is not a format

Reviewer 3: Also, did we really decide to use Wire as the default? I thought it would be Json.

## Using ModelSerializer for System.Text.Json

> In the following sample, we are using the default `ModelSerializerOptions`. This will allow the `ModelSerializer` to use the `System.Text.Json` serializer for all models. In the Deserialization sample, we are setting the `ModelSerializerFormat` in the Options to `Json`. This will serialize all properties including read-only and additional properties.

Reviewer 4: I don't think this description is accurate to what the code is doing?

This should be the very simple ModelSerializer case

Reviewer 3: Also, "read-only properties" is an Azure specific concept that has little to do with what users will think read-only property means. In general, I think this goes into too many details about the wire format. We should describe how the data/json format works, and then at the end mention that we can also produce wire format, how it differs, etc.

### Serialization

```C# Snippet:SystemTextJson_Serialize

> DogListProperty dog = new DogListProperty

Reviewer 4: Lets use a simpler model name like `Dog` or perhaps a real model from one of our services?  The ListProperty portion of this name is just to indicate a use case we are covering in our testing it doesn't mean anything for customers.

Reviewer 3: Could we use one of the real models in the SDK? As currently written, it's really not clear how DogListProperty is designed, why it works with the model serializer, etc. If we want to stick with DogListProperty we need to show how it's defined.

```C# Snippet:SystemTextJson_Serialize
DogListProperty dog = new DogListProperty
{
    Name = "Doggo",
    IsHungry = true,
    Weight = 1.1,
    FoodConsumed = { "kibble", "egg", "peanut butter" },
};
BinaryData data = ModelSerializer.Serialize(dog);
```

### Deserialization

```C# Snippet:ExplicitCast_Deserialize
PetStoreClient client = new PetStoreClient(new Uri("http://somewhere.com"), new MockCredential());
Response response = client.GetPet("myPet");
DogListProperty dog = (DogListProperty)response;
Console.WriteLine(dog.IsHungry);
```
```C# Snippet:SystemTextJson_Deserialize
ModelSerializerOptions options = new ModelSerializerOptions(ModelSerializerFormat.Json);
string json = @"[{""Name"":""Doggo"",""IsHungry"":true,""Weight"":1.1,""FoodConsumed"":[""kibble"",""egg"",""peanut butter""],""NumberOfLegs"":4}]";

Given that explicit cast does not allow for serialization options we might also consider a static `FromResponse` and instance `ToRequestContent` methods.
DogListProperty dog = ModelSerializer.Deserialize<DogListProperty>(BinaryData.FromString(json), options);
```

## Using ModelSerializer for NewtonSoftJson

> By using the ModelSerializer class, a new instance of Dog does not need to be created before calling Deserialize. Also added ObjectSerializer to Options class so different kinds of Serializers can be used.

Reviewer 4: I don't think this is one of the key concepts above, we might have a mismatch here?

> # Azure.Core Public Serialization Guide

Reviewer 2: ```suggestion
# Azure.Core public serialization
```

Our docs follow the [Microsoft style guide](https://learn.microsoft.com/en-us/style-guide/capitalization#sentence-style-capitalization-in-titles-and-headings) which calls for sentence-style capitalization in headers.

Reviewer 5: @nisha-bhatia This comment reminded me of something. You can run [Acrolinx in VS Code](https://aka.ms/azsdk/acrolinx-vscode) to check this doc against many of the style guide rules. It's licensed in our repo.

> The latest updates to the `Azure.Core` library have simplified serializing and deserializing of all public Azure models. With the addiion of the `ModelSerializer` class, we provide a public interface that exposes the previously internal serialization code so customers can access it without needing to use reflection or write their own translation. The recommended serialization approach is to use the static `ModelSerializer` class. The `ModelSerializer` class also has `ModelSerializerOptions` which allows the user to set the serialization `ModelSerializerFormat` type (`Json` or `Wire`) and provide custom Serializer types for specific models.

Reviewer 2: ```suggestion
The `Azure.Core` library supports serialization and deserialization of most client library model types using the static `ModelSerializer` class.  Using `ModelSerializerOptions` with the serializer allows you control over formatting and to override serialization behavior.  
```
- As a sample, we should be looking at this as a reference and not a point-in-time announcement of changes.  I'd suggest avoiding phrases like "the latest updates" and "with the addition" in favor of just discussing what the feature is and does. 

- I'd stay away from superlatives.  We can't guarantee ALL models now and forever.

- I'd suggest adding a dedicated section later in the sample that discusses the most useful options and what they mean.  Best to do that after the basic "how do I use this in the default case" sections.

Reviewer: +1 re: point in time.

> The latest updates to the `Azure.Core` library have simplified serializing and deserializing of all public Azure models. With the addiion of the `ModelSerializer` class, we provide a public interface that exposes the previously internal serialization code so customers can access it without needing to use reflection or write their own translation. The recommended serialization approach is to use the static `ModelSerializer` class. The `ModelSerializer` class also has `ModelSerializerOptions` which allows the user to set the serialization `ModelSerializerFormat` type (`Json` or `Wire`) and provide custom Serializer types for specific models.

Reviewer: In the intro, I'd love to see something like:

Problem framing:
- .NET serialization typically uses JsonSerializer in System.Text.Json
- Azure SDK models can't use that because ___
- That means, for Azure SDK models we need to do _____
- IModelSerializable solves this problem

Overview of serialization types:
- Azure SDK models implement IModelSerializer (because we said that above)
- This can be hard to use as such because ___
- To make it easier to serialize Azure SDK models, we provide ModelSerializer, because it _______ (knows how to use these interfaces?  lets us provide a set of options?  gives us good performance?)

After that, key concepts can follow like:
- What if my models use XML?  IXmlModelSerializable
- How can this help me use protocol methods?  DynamicData
- Well, what if I have a model that's super customized and needs Newtonsoft.JSON?  We have "bring your own model"

In general, each of our types should be addressing a business need - we can help customers see how they can use them by telling them why each thing exists.

## Key Concepts

> - [ModelSerializerFormat description]("modelserializerformat-description")

Reviewer 2: This does not exist.  

When adding it, I'd suggest that you place it after the basic "how to use this" sections.   Our goal is to let folks who want to scan and get started quickly find the "how to" content and then come back for the "how do i?" and "how does this work" content.

- [Using the ModelSerializer](#using-the-modelserializer)
- [Interfacing with protocol methods](#interfacing-with-protocol-methods)
- [Using JsonSerializer](#using-jsonserializer)
- [Envelope bring your own model case](#envelope-bring-your-own-model-case)
- [Interfacing with the low level interfaces](#interfacing-with-the-low-level-interfaces)

## Using the ModelSerializer

> The default serialization options can be overridden by passing in a `ModelSerializerOptions` object to the Serialize and Deserialize methods. Developers can set the format options to specify the serialization format such as XML or JSON. Default `ModelSerializationOptions` use the `Json` format and the `System.Text.Json` serializer. The following samples demonstrate how to use `ModelSerializer` class for `System.Text.Json` and `Newtonsoft.Json` serialization.

Reviewer 2: This seems like a less mainstream and more advanced use case.   Let's start with the basic "how to" guide using default options for serialization and deserialization.  

As written, there's a paragraph of information that I need to absorb on options for different formats and serialization backends that I have to get through before I get to see the "Hello world" examples.    

Let's get through the basics before talking about the ways that I can customize.

### Using ModelSerializer for System.Text.Json

> In the following sample, we are using the default `ModelSerializerOptions`. This will allow the `ModelSerializer` to use the `System.Text.Json` serializer for all models. In the Deserialization sample, we are setting the `ModelSerializerFormat` in the Options to `Json`. This will serialize all properties including read-only and additional properties.

Reviewer 2: Let's skip calling them the "default options" and just ignore that we can override things for the "Hello world" case.

### Serialization

> ```C# Snippet:SystemTextJson_Serialize

Reviewer 2: Since they must be globally unique, let's please use snippet names that are explicitly scoped to Azure.Core.   This form is quite general and may cause collisions with other libraries.

DogListProperty dog = new DogListProperty
{
    Name = "Doggo",
    IsHungry = true,
    Weight = 1.1,
    FoodConsumed = { "kibble", "egg", "peanut butter" },
};
BinaryData data = ModelSerializer.Serialize(dog);
```

### Deserialization

```C# Snippet:SystemTextJson_Deserialize
ModelSerializerOptions options = new ModelSerializerOptions(ModelSerializerFormat.Json);

> string json = @"[{""Name"":""Doggo"",""IsHungry"":true,""Weight"":1.1,""FoodConsumed"":[""kibble"",""egg"",""peanut butter""],""NumberOfLegs"":4}]";

Reviewer 2: We should pretty format JSON literals like this.  Our goal should be to help developers visualize the structure of the JSON at a glance and mentally map it to what they would expect the model type to look like.

DogListProperty dog = ModelSerializer.Deserialize<DogListProperty>(BinaryData.FromString(json), options);
```

### Using ModelSerializer for NewtonSoftJson

> In the following sample, we are adding the `DogListProperty` with the `NewtonsoftJsonObjectSerializer` to the `GenericTypeSerializerCreator`. This will allow the `ModelSerializer` to use the `NewtonsoftJsonObjectSerializer` for the DogListProperty model. 

Reviewer 2: This text describes what the code is doing - which I can see by reading the code.   What I think we want to focus on is explaining "why" for anything non-obvious.   

In this case, for example, I have no idea what the heck a `GenericTypeSerializerCreator` is - the name means nothing to me.   It would be helpful to explain what that is used for any why we need to set that to use Newtonsoft.

DogListProperty dog = ModelSerializer.Deserialize<DogListProperty>(BinaryData.FromString(json), options);
```

## Using explicit cast

> If you would like to convert the protocol model to the strongly typed model, you can use the explicit cast operator. This will allow you to use the strongly typed model for serialization and deserialization. There is also an explicit cast operator that can be used to convert the strongly typed model to the protocol RequestContent.

Reviewer 3: "protocol model" is not defined.

### Serialization

```C# Snippet:ExplicitCast_Serialize
DefaultAzureCredential credential = new DefaultAzureCredential();
PetStoreClient client = new PetStoreClient(new Uri("http://somewhere.com"), credential);
DogListProperty dog = new DogListProperty("myPet");

> Response response = client.CreatePet("myPet", (RequestContent)dog);

Author: KrzysztofCwalina 

I would add a comment explaining why one line used a cast and the other uses Create

Response response2 = client.CreatePet("myPet", RequestContent.Create(dog));
```

### Deserialization

```C# Snippet:ExplicitCast_Deserialize
DefaultAzureCredential credential = new DefaultAzureCredential();
PetStoreClient client = new PetStoreClient(new Uri("http://somewhere.com"), credential);
Response response = client.GetPet("myPet");
DogListProperty dog = (DogListProperty)response;
Console.WriteLine(dog.IsHungry);
```

## Using ModelJsonConverter for JsonSerializer

If you have Json that needs to be converted into a specific object type, consider using the `ModelJsonConverter class`. This class can handle the deserialization of a model to a specific type and include additional metadata. 

> ### Serialization

Author: heaths 

Lots of programmatic constructs herein. Elsewhere. Just wanted to point out again my comment above applies to any markdown in this PR. We should be doing this consistently.

DogListProperty dog = ModelSerializer.Deserialize<DogListProperty>(BinaryData.FromString(json), options);
```

> ## Interfacing with protocol methods

Reviewer 2: We're making an assumption that developers know what a "protocol method" means to us.   While we use this internally, it's not an industry-standard term and is unlikely to be meaningful to developers using our libraries.

We should define the term, offering a short sentence or three on what a "protocol method" is.  We should also consider a title that doesn't use a non-obvious term.  Maybe something like "Interfacing with low-level methods" or "Working with HTTP methods" or... uh, something.

If you would like to convert the protocol model to the strongly typed model, you can use the explicit cast operator. This will allow you to use the strongly typed model for serialization and deserialization. There is also an explicit cast operator that can be used to convert the strongly typed model to the protocol RequestContent.

### Serialization

```C# Snippet:ExplicitCast_Serialize
DefaultAzureCredential credential = new DefaultAzureCredential();
PetStoreClient client = new PetStoreClient(new Uri("http://somewhere.com"), credential);
DogListProperty dog = new DogListProperty("myPet");

> Response response = client.CreatePet("myPet", (RequestContent)dog);

Reviewer 4: This cast should be implicit we can remove the `(RequestContent)`

DogListProperty dog = System.Text.Json.JsonSerializer.Deserialize<DogListProperty>(json, options);
```

> ## Envelope bring your own model case

Reviewer 2: I've got no idea what this title means.   The core scenario here seems useful - how to use a converter to customize things.   We should consider revising the title so that it describes the scenario.

Reviewer 6: I don't think `Envelope` or `BYOM` mean anything to customers.  I'd change this to just `Generic Models`.

Envelope<ModelT> model = ModelSerializer.Deserialize<Envelope<ModelT>>(new BinaryData(Encoding.UTF8.GetBytes(serviceResponse)), options: options);
```

> ## Interfacing with the low level interfaces

Reviewer 2: ```suggestion
## Interfacing with the low-level interfaces
```

I'd suggest that we refine this title.   "Low-level interfaces" is a subjective term and doesn't really tell me what they are - just what we call them internally.   To be honest, I wanted to make suggestions here for alternatives, but I have no idea what the heck these are.

In this example we demonstrate how we could use the exact same method / interface to serialize either Json or Xml.
Above we demonstrate using `SerializeJson` and `SerializeXml` methods but this requires the user to know which serializer to use.

> ### Serialization

Reviewer 2: I find the fact that we're creating an instance of something called "interface" incredibly confusing.  The term has a very specific meaning to .NET developers - and we're using it in a very different way.   I don't know if the ship has sailed on naming at this point, but I'd suggest we reconsider this one.

//cc: @m-nash, @KrzysztofCwalina

string jsonResponse = "{\"key\":\"Color\",\"value\":\"Red\",\"readOnlyProperty\":\"ReadOnly\",\"x\":\"extra\"}";
JsonModelForCombinedInterface jsonModel = ModelSerializer.Deserialize<JsonModelForCombinedInterface>(new BinaryData(Encoding.UTF8.GetBytes(jsonResponse)));
```

## Next steps

> To learn more about serialization with Azure Core, see the [Azure.Core.Serialization class](https://learn.microsoft.com/en-us/dotnet/api/azure.core.serialization.objectserializer?view=azure-dotnet).

Reviewer 2: ```suggestion
To learn more about serialization with `Azure.Core`, see the [Azure.Core.Serialization namespace](https://learn.microsoft.com/dotnet/api/azure.core.serialization.objectserializer?view=azure-dotnet).
```

> ## Contributing

Reviewer 2: This doesn't appear in any other sample for any library.  Please remove.
